source/mir/ndslice/algorithm.d

```diff
@@ -1884,3 +1885,714 @@ size_t countImpl(alias fun, Slices...)(Slices slices)
     while(!slices[0].empty);
     return ret;
 }
+
+version(unittest)
+{
+    T identity(T)(T x)
+    {
+        return x;
+    }
+}
+
+/++
+The call `eachLower!(fun)(matrix)` applies `fun` to the lower triangle of
+a two-dimensional slice.
+
+The value `k` determines which diagonals will have the function applied:
+For k = 0, the function is also applied to the main diagonal
+For k = 1 (default), only the non-main diagonals below the main diagonal
+will have the function applied.
+For k > 1, fewer diagonals below the main diagonal will have the function
+applied.
+For k < 0, more diagonals above the main diagonal will have the function
+applied.
+
+Params:
+    fun = A function
+    k = adjustment to diagonals (default = 1)
++/
+template eachLower(alias fun, ptrdiff_t k = 1)
+{
+    import mir.functional : naryFun;
+    import mir.ndslice.algorithm : eachImpl;
+    import mir.ndslice.slice : Slice, SliceKind;
+
+    static if (__traits(isSame, naryFun!fun, fun))
+    {
+        void eachLower(SliceKind kind, Iterator)
+                           (Slice!(kind, [2], Iterator) matrix)
+        {
+            immutable(size_t) m = matrix.length!0;
+            immutable(size_t) n = matrix.length!1;
+
+            void eachLowerImpl(T)(T matrix)
+            {
+                void eachLowerImpl_i(U)(U e, size_t i)
+                {
+                    if (i < (n + k))
+                        e[0..(i - k + 1)].eachImpl!fun;
+                    else
+                        e.eachImpl!fun;
```

Comment:
`eachImpl` probably will be inlined, so the following code will be two times smaller (in terms of object code size):
```d
eachImpl!fun(i < n + k ? e[0..(i - k + 1)] : e);
```

EDIT: generally we should not have conditions in the loop. So dismiss this comment.